### ISSUE TITLE
Server pro: redirection after logging in is broken when using RStudio's documentation for running server behind proxy with custom path

### ISSUE THREAD

**Issue by daattali** [rstudio/shiny-server#387]
## Summary

I installed shiny server commercial v 1.5.10.990 on ubuntu 16.04 and followed the instructions [here](https://support.rstudio.com/hc/en-us/articles/213733868-Running-Shiny-Server-with-a-Proxy) which were updated one month ago. I built a test app that requires simple flatfile authentication. After logging in, an extra `/shiny` gets inserted in the URL.

## Demo

Go to http://104.248.71.37/shiny/test/ and log in using the credentials `test`/`shiny1`. You will be redirected to http://104.248.71.37/shiny/shiny/test/ which of course will not work. You need to manually remove `shiny/` from the URL and now you're in. If you use http://104.248.71.37:3838/test/ instead (using the port instead of /shiny) then it does work as expected.

## Steps to reproduce

1. Install nginx, R, the `shiny` package, gdebi-core , and shiny-server-commercial-1.5.10.990-amd64.deb

2. Create the following test app in `/srv/shiny-server/test/app.R`:

    ```
    ui <- fluidPage("hello", textOutput("user"))
    server <- function(input, output, session) {
        output$user <- renderText(session$user)
    }
    shinyApp(ui, server)
    ```

3. Use the following shiny server configuration `/etc/shiny-server/shiny-server.conf`

    ```
    run_as shiny;
    
    server {
      listen 3838;
    
      location / {
        site_dir /srv/shiny-server;
        log_dir /var/log/shiny-server;
        directory_index on;
    
        location /test {
          required_user *;
        }
      }
    }
    
    auth_passwd_file /etc/shiny-server/passwd;
    ```

  (then restart the shiny server)

4. Create a test user using the command `sudo /opt/shiny-server/bin/sspasswd /etc/shiny-server/passwd test`

5. Use the following nginx config file, which is what the link above suggests when using a custom path with a proxy `/etc/nginx/sites-enabled/default`

    ```
    map $http_upgrade $connection_upgrade {
            default upgrade;
            ''      close;
    }
    
    server {
            listen 80 default_server;
            listen [::]:80 default_server;
    
            root /var/www/html;
    
            index index.html index.htm index.nginx-debian.html;
    
            server_name _;
    
            location / {
                    try_files $uri $uri/ =404;
            }
    
            rewrite ^/shiny$ $scheme://$http_host/shiny/ permanent;
    
            location /shiny/ {
                    rewrite ^/shiny/(.*)$ /$1 break;
                    proxy_pass http://localhost:3838;
                    proxy_redirect / $scheme://$http_host/shiny/;
                    proxy_http_version 1.1;
                    proxy_set_header Upgrade $http_upgrade;
                    proxy_set_header Connection $connection_upgrade;
                    proxy_read_timeout 20d;
                    proxy_buffering off;
            }
    }
    ```

    (then restart nginx)

6. Go to `mysite.com/shiny/test` and login. You will be redirected to `mysite.com/shiny/shiny/test`.

## Solution

In a [very old article](https://deanattali.com/2015/05/09/setup-rstudio-shiny-server-digital-ocean/) I came up with an nginx configuration that seemed to have withstand the test of time. I'm not sure if it's 100% correct and safe, but I will a simplified version here because maybe it's useful, since it seems to not have the above problem:

```
map $http_upgrade $connection_upgrade {
        default upgrade;
        ''      close;
}

server {
        listen 80 default_server;
        listen [::]:80 default_server;

        root /var/www/html;

        index index.html index.htm index.nginx-debian.html;

        server_name _;

        location / {
                try_files $uri $uri/ =404;
        }

        location /shiny/ {
                rewrite ^(/shiny/[^/]+)$ $1/ permanent;
                proxy_pass http://127.0.0.1:3838/;
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection $connection_upgrade;
        }
}
```

**Comment by daattali** [comment]
I'm killing the server that was running the demo, but the steps to reproduce are simple

**Comment by jcheng5** [comment]
@colearendt Any thoughts?

**Comment by colearendt** [comment]
Yeah, I would actually call this inconsistent behavior by Shiny Server, and I think it surfaces now because we relatively recently fixed the Support article, which did not handle redirects properly. Shiny Server almost universally issues relative redirects (i.e. if you request `/path`, it will redirect you to `/path/`). 

In this case, though, the request body for a successful has user / pass / success_url in it.

<img width="351" alt="image" src="https://user-images.githubusercontent.com/23075542/52023425-17eed580-24cb-11e9-99cd-7d8842bb6a96.png">

Shiny Server seems to use this success_url to issue the redirect for "where to go after login." Hence it disregards the "request URL", which the proxy will rewrite from `http://localhost:32772/shiny/test/__login__` to `http://localhost:32772/shiny/__login__` (and which thus has no effect).

TL;DR; Shiny Server pro tells our user to go to `/shiny/test`, which does not exist from the Shiny Server's perspective. The proxy is actually handling this behavior _correctly_ in my opinion, because it thinks "Oh, Shiny Server wants you to go to `/shiny/test`, that is actually at `/shiny/shiny/test`"

The only workaround I have come up with (which I did a sanity check on but have not tested "thoroughly") is to add another `proxy_redirect` line to take care of this case. Note that this would cause conflicts and problems if any assets are hosted at `/shiny` from the Shiny Server's perspective - I would like to think this is unlikely because then the user would be typing `/shiny/shiny` and nobody likes that. 

```
      proxy_redirect /shiny/ $scheme://$http_host/shiny/;

      # in addition to the existing
      proxy_redirect / $scheme://$http_host/shiny/;
```

@jcheng5 Do you think this behavior could be modified easily in Shiny Server Pro, or can you think of another workaround? Is it possible for nginx to edit the request body? 😱 That seems like undesirable behavior.

**Comment by colearendt** [comment]
It looks like @daattali 's solution just doesn't handle redirects at all, but works around the most common case (lack of a trailing slash) by rewriting those at the proxy with `rewrite ^(/shiny/[^/]+)$ $1/ permanent;` Definitely workable, but personally I would prefer that Shiny Server issue redirects consistently in case there is ever a case where Shiny Server needs to issue a redirect that is more than just adding a trailing slash.

**Comment by daattali** [comment]
@colearendt can you show me what case is not handled by my config? I'm surprised nobody ever told me that my config didn't work for them, so I'm just curious what case would cause it to break!

**Comment by colearendt** [comment]
Right, I was trying to think of a case where it would fail too (i.e. the question above). I think it might require that you are accessing the proxy on a port other than what nginx is listening on, and then Shiny Server issues a redirect other than adding a trailing slash. I.e. nginx is listening on port 80, but it is running in a docker container that is actually mapped to port 8080 on the host.

That is what the `$http_host` variable is supposed to remedy. The trick, though, is that I'm not sure if Shiny Server ever issues such a redirect... 🤷‍♂️ So it may be a purely theoretical fear 😆 

Also, it may be a nonissue because all of Shiny Server's redirects are relative, and it seems you are more wholly embracing that approach whereas the Support article is attempting to make absolute redirects. So maybe that wouldn't be a problem after all!

**Comment by colearendt** [comment]
Oops. Another case yours _might_ break. It looks like you add a trailing `/` to basically everything (presuming on Shiny Server's behavior that just about everything is a directory), so you could potentially have problems with static files or redirects that do not go to folders (where a trailing slash would be unnecessary or invalid).

Again, I might be crazy. Reasoning about proxies makes me 🤕
